### DISCUSSION TITLE
Catching maximum message size exceeded

### DISCUSSION THREAD

**Discussion by daleif** [mojolicious/mojo#2208] [General]
Just had issue with an "The connection was reset" problem due to trying to upload a file that was a bit too big.

Under morbo I can see I get `[trace] Maximum message size exceeded` is there a hook one can use to catch these so we can inform the user properly instead of getting these connection was reset errors?

So far I haven't figured out what is the best approach here. I'd also like to inform webmaster when these happen. But it seems I need to catch them quite early.

**Comment by kraih** [comment]
I believe that is still not possible.

**Reply by daleif** [reply]
Dang.

Out of curiosity, where in the process is the request being rejected? 

I just find it interesting that my code for evaluating the form still runs even if the request is too large.

**Reply by kraih** [reply]
Look for `Maximum message size exceeded` in the code. Not that it matters much, it's the network protocols that limit what we can do.

**Reply by daleif** [reply]
It seems to be coming directly from Mojo::Message, the `parse` routine runs

```
  # Check message size
  my $max = $self->max_message_size;
  return $self->_limit('Maximum message size exceeded') if $max && $max < $self->{raw_size};
```
with

```
sub _limit { ++$_[0]{limit} and return $_[0]->error({message => $_[1]}) }
```

So not much we can do

**Reply by guest20** [reply]
You can set the limit to (including to zero to disable checking the limit) in your app's `startup`...

**Reply by kraih** [reply]
Maybe i wasn't clear enough, the limit enforcement is completely irrelevant because browsers do not check for a response before the entire request has been sent. You either have a limit, or you always receive the entire request and then manually reject it for being too large. Which seems rather pointless, since you've already retrieved everything anyway.

**Reply by daleif** [reply]
> You can set the limit to (including to zero to disable checking the limit) in your app's `startup`...

I already set a larger upper limit in this case.

**Reply by daleif** [reply]
> Maybe i wasn't clear enough, the limit enforcement is completely irrelevant because browsers do not check for a response before the entire request has been sent. You either have a limit, or you always receive the entire request and then manually reject it for being too large. Which seems rather pointless, since you've already retrieved everything anyway.

I can see why the max setting is there. 

The point here is the unexpected situation where someone uploaded something much larger. It took me quite a while to figure out what the problem was. So just wanted to either log that this was rejected due to size, or send a message to webmaster. When running under morbo the message from Mojo::Message was in the log as a trace, but in production it of course was never seen.

My wondering was just where in the process the limit is enforced.

**Reply by s1037989** [reply]
@daleif Can you subscribe to the Mojo::Message error event and write your own error log entry?
It sounds like what you want is `$self->_limit('Maximum message size exceeded')` to be logged to `error` instead of `trace`.  Is this error (the one called in `sub _limit {...}`) emitted to the `error` event handler?

**Reply by daleif** [reply]
@s1037989  stupid question, as I hardly ever use the various events. 

Where exactly would one subscribe to that error event? It is probably some `$self->on('error'....)` somewhere. But it keeps giving me errors or does not do anything. (one of the reasons why I tend to no use them, they are a bit confusing.)

**Reply by s1037989** [reply]
@daleif Not stupid at all as I'm just grasping for straws... I've never used it in your situation, but since Mojo::Message is a sub class of Mojo::EventEmitter which has an error event, I thought maybe if you attached something like a `$c->req->on(error => sub {...})` event handler maybe it'll pop and get something to allow you to throw a log?  But I did see in the code that `sub _limit {}` doesn't explicitly `$self->emit('error')` or `die` or `Mojo::Exception->throw`...  Anyway, what you want, I reckon, is just some event or hook that will give you the opportunity to add an additional `$log->error` to the `$log->trace` that's already called by `sub _limit {}`.  I think you were generally looking for a `hook` solution which I think would generally require detecting the problem by analyzing state at the time of the hook, whereas events are thrown in response to certain conditions and no deep state inspection is usually needed.  EventEmitter is probably one of the most power mechanisms in Mojo so don't give up on them!

**Reply by s1037989** [reply]
@daleif Sorry for the misleading suggestion...  doesn't look good.  I modified [request.t L25](https://github.com/mojolicious/mojo/blob/main/t/mojo/request.t#L25) to add `$req->on(error => sub { warn 'ERROR!' });` but it never popped (unlike `$req->on(finish => sub {})`).
So what about a `before_dispatch` hook that looks at `$c->req->error->{message}`?

**Reply by s1037989** [reply]
Looks like **hook** might be the right answer!

```
$ env MOJO_LOG_LEVEL=info perl -Mojo -E 'app->max_request_size(100); hook before_dispatch => sub ($c) { $c->log->error($c->req->error->{message}) if $c->req->error }; a("/" => {text => "ok"}); app->start' get /
ok

$ env MOJO_LOG_LEVEL=info perl -Mojo -E 'app->max_request_size(1); hook before_dispatch => sub ($c) { $c->log->error($c->req->error->{message}) if $c->req->error }; a("/" => {text => "ok"}); app->start' get /
[2024-11-21 20:45:53.42485] [2433908] [error] [OnoozdbjB4_m] Maximum message size exceeded
```

**Reply by daleif** [reply]
I'll give that a go. Is already using `before_render` to send crashes directly to my inbox. Not sure if `before_render` is the right place for that, but it works. I'll do some testing with `before_dispatch` to see if it works as intended.
